linalg/ode.hpp

```diff
@@ -12,13 +12,91 @@
 #ifndef MFEM_ODE
 #define MFEM_ODE
 
+#include "../general/communication.hpp"
 #include "../config/config.hpp"
 #include "operator.hpp"
-#include "../general/communication.hpp"
+#include <vector>
+#include <memory>
 
 namespace mfem
 {
 
+/// An interface for storing state of previous timesteps
+class ODEStateData
+{
+public:
+   /// Get the maximum number of stored stages
+   virtual int MaxSize() const = 0;
+
+   /// Get the current number of stored stages
+   virtual int Size() const = 0;
+
+   /// Get the ith state vector
+   virtual void Get(int i, Vector &state) const = 0;
+
+   /// Get the ith state vector
+   virtual const Vector &Get(int i) const = 0;
+
+   /// Set the ith state vector
+   virtual void Set(int i, Vector &state) = 0;
+
+   /// Add state vector and increment state size
+   virtual void Add(Vector &state) = 0;
+
+   /// Virtual destructor
+   virtual ~ODEStateData() = default;
+};
+
+/// An implementation of ODEStateData that stores states in an std::vector<Vector>
+class ODEStateDataVector : public ODEStateData
+{
+private:
+   MemoryType mem_type;
+   int ss, smax;
+   std::vector<Vector> data;
+   Array<int> idx;
+
+public:
+   ODEStateDataVector () { ss = smax = 0;};
+
+   /// Set the number of stages and the size of the vectors
+   void  SetSize(int stages, int vsize, MemoryType mem_type);
+
+   /// Shift the stage counter for the next timestep
+   inline void ShiftStages()
+   {
+      for (int i = 0; i < smax; i++) { idx[i] = (++idx[i])%smax; }
+   };
+
+   /// Increment the stage counter
+   void Increment() { ss++; ss = std::min(ss,smax); };
+
+   /// Reset the stage counter
+   void Reset() { ss = 0; };
+
+   /// Reference access to the ith vector.
+   inline Vector & operator[](int i) { return data[idx[i]]; };
+
+   /// Const reference access to the ith vector.
+   inline const Vector &operator[](int i) const { return data[idx[i]]; };
+
+   /// Print state data
+   void Print(std::ostream &out = mfem::out) const ;
```

Comment:
Use `os` instead of `out` to avoid confusion with `mfem::out`. This is already done in the implementation.
```suggestion
   void Print(std::ostream &os = mfem::out) const ;
```

Reply:
Done.

Reply:
Done